File: server/src/routes/ingest.ts
import { Router } from 'express';
import multer from 'multer';
import { parse } from 'csv-parse';
import { z } from 'zod';
import { prisma } from '../shared/prisma.js';

const router = Router();
const upload = multer({ storage: multer.memoryStorage(), limits: { fileSize: 5 * 1024 * 1024 } });

const rowSchema = z.object({
  course_id: z.string().min(1),
  course_name: z.string().min(1),
  department: z.string().min(1),
  level: z.enum(['UG', 'PG']),
  delivery_mode: z.enum(['online', 'offline', 'hybrid']),
  credits: z.coerce.number().int().min(0),
  duration_weeks: z.coerce.number().int().min(1),
  rating: z.coerce.number().min(0).max(5),
  tuition_fee_inr: z.coerce.number().min(0),
  year_offered: z.coerce.number().int(),
});

router.post('/', upload.single('file'), async (req, res) => {
  const token = req.headers['x-ingest-token'] as string | undefined;
  if (!token || token !== process.env.INGEST_TOKEN) {
    return res.status(401).json({ error: 'Unauthorized' });
  }
  if (!req.file) {
    return res.status(400).json({ error: 'CSV file is required under field name "file"' });
  }
  const text = req.file.buffer.toString('utf8');

  const records: any[] = [];
  await new Promise<void>((resolve, reject) => {
    const parser = parse(text, { columns: true, trim: true, skip_empty_lines: true });
    parser.on('readable', () => {
      let record;
      while ((record = parser.read())) {
        const parsed = rowSchema.safeParse(record);
        if (!parsed.success) {
          reject({ error: 'Invalid row', details: parsed.error.flatten(), record });
          return;
        }
        records.push(parsed.data);
      }
    });
    parser.on('error', (err) => reject(err));
    parser.on('end', () => resolve());
  });

  // Upsert rows
  const operations = records.map((r) =>
    prisma.course.upsert({
      where: { course_id: r.course_id },
      update: r,
      create: r,
    })
  );
  await prisma.$transaction(operations);

  res.json({ inserted: records.length });
});

export default router;

File: backend/routes/ingest.py
from flask import Blueprint, request
from werkzeug.utils import secure_filename
from models.course import Course
from db import db
import csv
import io
import os

ingest_bp = Blueprint('ingest', __name__)

@ingest_bp.post('/')
def ingest():
    token = request.headers.get('x-ingest-token')
    if not token or token != os.getenv('INGEST_TOKEN'):
        return {'error': 'Unauthorized'}, 401

    if 'file' not in request.files:
        return {'error': 'CSV file required in field "file"'}, 400

    f = request.files['file']
    filename = secure_filename(f.filename)
    if not filename.lower().endswith('.csv'):
        return {'error': 'Only CSV files are accepted'}, 400

    content = f.read().decode('utf-8')
    reader = csv.DictReader(io.StringIO(content))

    required = ['course_id','course_name','department','level','delivery_mode','credits','duration_weeks','rating','tuition_fee_inr','year_offered']
    for col in required:
        if col not in reader.fieldnames:
            return {'error': f'Missing column {col}'}, 400

    count = 0
    for row in reader:
        try:
            course = Course(
                course_id=row['course_id'].strip(),
                course_name=row['course_name'].strip(),
                department=row['department'].strip(),
                level=row['level'].strip(),
                delivery_mode=row['delivery_mode'].strip(),
                credits=int(row['credits']),
                duration_weeks=int(row['duration_weeks']),
                rating=float(row['rating']),
                tuition_fee_inr=int(row['tuition_fee_inr']),
                year_offered=int(row['year_offered']),
            )
        except Exception as e:
            return {'error': 'Invalid row', 'details': str(e), 'row': row}, 400

        existing = Course.query.get(course.course_id)
        if existing:
            # Update existing
            for k, v in course.to_dict().items():
                setattr(existing, k, v)
        else:
            db.session.add(course)
        count += 1

    db.session.commit()
    return {'inserted': count}